### PR TITLE
Feat/expire peers

### PR DESCRIPTION
**Pull request by conectado**
This PR takes care of expiring connections with peer from the gateway side.

### CODE REVIEW

File: rust/connlib/libs/tunnel/src/peer.rs [hunk 1]
```diff
@@ -49,13 +59,26 @@ impl Peer {
...
+    pub(crate) async fn shutdown(&self) -> Result<()> {
+        self.channel.close().await?;
+        Ok(())
+    }
+
+    pub(crate) fn is_valid(&self) -> bool {
+        match self.expires_at {
+            Some(expires_at) => expires_at > Utc::now(),
+            None => true,
+        }
```
Comment (francesca64) [inline rust/connlib/libs/tunnel/src/peer.rs]:
```suggestion
        !self.expires_at.is_some_and(|expires_at| expires_at <= Utc::now())
```

File: rust/Cargo.toml [hunk 2]
```diff
@@ -15,6 +15,7 @@ members = [
 [workspace.dependencies]
 boringtun = { git = "https://github.com/firezone/boringtun", branch = "master", default-features = false }
 swift-bridge = { git = "https://github.com/chinedufn/swift-bridge.git", rev = "4fbd30f" }
+chrono = { version = "0.4", default-features = false, features = ["std", "clock", "oldtime", "serde"] }
```
Comment (francesca64) [inline rust/Cargo.toml:18]: I'm curious what the rationale is for disabling `default-features` and then manually specifying the default features anyway? I've noticed `default-features = false` is common throughout our dependency specifications, so I assume it's a policy we have?

**Review by francesca64 — CHANGES_REQUESTED** [review]

File: rust/Cargo.toml (hunk 2 above)
Comment (conectado) [reply rust/Cargo.toml:18]: It's an undocumented convention I have, I found it makes it easier to track features long-term specially if we eventually wanted to move to some no-std environment and generally I prefer to be explicit about the features we use.

I don't feel too strongly about this, if you think it's better we can just use default features.

Comment (francesca64) [reply rust/Cargo.toml:18]: I definitely understand wanting to make features less opaque, but this adds substantial tedium/noise and is rather unconventional. In the long-term that will be a stumbling block for open source contributors.

I'm guessing `cargo add` can't be configured to do this automatically? Do we have any concrete plans to support no-std?

Comment (conectado) [reply rust/Cargo.toml:18]: > I'm guessing cargo add can't be configured to do this automatically? 

Not AFAIK

> Do we have any concrete plans to support no-std?

Nothing really concrete

Comment (francesca64) [reply rust/Cargo.toml:18]: I don't think the benefits outweigh the surprise then.

Comment (conectado) [reply rust/Cargo.toml:18]: Alright, will change this in a separate PR

File: rust/connlib/libs/tunnel/src/control_protocol.rs [hunk 3]
```diff
@@ -285,15 +286,28 @@ where
...
                             tunnel.callbacks.on_error(&e, Recoverable);
                             tracing::error!(
                                 "Couldn't establish wireguard link after opening channel: {e}"
                             );
                             // Note: handle_channel_open can only error out before insert to peers_by_ip
                             // otherwise we would need to clean that up too!
-                            tunnel.peer_connections.lock().remove(&client_id);
+                            let conn = tunnel.peer_connections.lock().remove(&client_id);
+                            if let Some(conn) = conn {
+                                // We are already on an error state, we ignore if the close errors also
```
Comment (francesca64) [inline rust/connlib/libs/tunnel/src/control_protocol.rs]: Wouldn't it still be good to log any errors?

**Review by francesca64 — APPROVED** [review]

**Pull request merged**